File: Golang/First-Missing-Positive.go
/*
  Time: O(n)
  Space: O(n)
*/
func firstMissingPositive(nums []int) int {
    maxNum := nums[0]
    
    num := map[int]int{}
    
    for i := 0; i<len(nums); i++ {
        if nums[i] > maxNum {
            maxNum = nums[i];
        }
        num[nums[i]] = nums[i]
    }
    
    if maxNum < 1 {
        return 1
    }
    
    for i := 1; i<maxNum+2; i++ {
        if _, exists := num[i]; !exists {
            return i
        }
    }
    return -1
}

File: Python/Algorithms/First-Missing-Positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        maxNum = nums[0]
        
        # store elements of nums in a dict for constant lookup
        num = {i:i for i in nums}
        
        for i in nums:
            maxNum = max(maxNum, i)
        
        if maxNum < 1:
            return 1
        for i in range(1, maxNum+2):
            if i not in num:
                return i
        
            
            
